[FEAT] 회원 정보 수정 API 구현

dye0p opened the pull request:
## Issue Number
closed #1048 

## As-Is
<!-- 문제 상황 정의 -->
- 현재 사용자는 자신의 회원 정보를 조회할 수만 있고, 수정하는 기능은 제공되지 않습니다.

## To-Be
<!-- 변경 사항 -->
- 로그인한 사용자가 자신의 이름, 성별, 비밀번호, 전화번호를 수정할 수 있는 API를 추가했습니다.
- `PATCH /members/me` 엔드포인트를 통해 요청하며, 선택적인 정보만 포함하여 수정할 수 있습니다.
- 수정되지 않은 필드는 요청 dto에 null로 들어옵니다.
- 수정에 성공하면 상태코드 `204 NO_CONTENT`를 응답합니다.
- 기존 문자열로 받던 성별을 Enum화 하였습니다. (MALE/FEMALE)


## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description
- `MemberInfoUpdateRequest` DTO를 추가하여 요청 데이터를 처리합니다.
- 기능 추가에 따라 `MemberServiceTest`와 `MemberIntegrationTest`에 테스트 코드를 작성했습니다.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## 릴리즈 노트

* **새로운 기능**
  * 회원 정보 업데이트 기능 추가: 사용자가 이름, 성별, 휴대폰번호, 비밀번호를 수정할 수 있습니다.

* **API 개선**
  * 회원 관련 API 엔드포인트 구조 정리 및 최적화

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dye0p committed:
refactor: final 키워드 순서 변경

dye0p committed:
feat: 회원 정보 수정 요청 dto 추가

dye0p committed:
feat: 회원 수정 API 추가

dye0p committed:
refactor: 공통 prefix RequestMapping 선언

dye0p committed:
test: 회원 정보 수정 통합 테스트 추가

dye0p committed:
test: 회원 정보 수정 API 테스트 추가

dye0p committed:
refactor: MemberService 중복 예외 처리 로직 제거

dye0p committed:
feat: 회원정보 수정 요청 dto 검증 추가

- @Valid 어노테이션 추가로 요청 값 유효성 검증 적용
- MemberController의 updateInfo 메서드에 @Valid 적용

dye0p commented:
@CodeRabbit 리뷰해

kimdevv commented on backend/fittoring/src/main/java/fittoring/application/member/service/MemberService.java:
```diff
@@ -49,18 +50,40 @@ private Image findMentoringImage(Mentoring mentoring) {
-                .orElseThrow(() -> new NotFoundMemberException(BusinessErrorMessage.LOGIN_ID_NOT_FOUND.getMessage()));
+        Member member = getMember(memberId);
         if (member.isNotAdmin()) {
             throw new ForbiddenException(BusinessErrorMessage.FORBIDDEN_MEMBER.getMessage());
         }
         return true;
     }
 
+    @Transactional
+    public void updateMemberInfo(Long memberId, MemberInfoUpdateRequest request) {
+        Member member = getMember(memberId);
+        if (request.name() != null) {
+            member.updateName(request.name());
+        }
+
+        if (request.gender() != null) {
+            member.updateGender(request.gender());
+        }
+
+        if (request.phoneNumber() != null) {
+            member.updatePhoneNumber(request.phoneNumber());
+        }
+
+        if (request.password() != null) {
+            member.updatePassword(request.password());
+        }
+    }
```
p1: 토끼의 멘트 좋은 것 같아요 특히 2번을 수정하지 않으면 오류를 발생시킬 수도 있겠네요~

kimdevv commented on backend/fittoring/src/main/java/fittoring/domain/model/Member.java:
```diff
@@ -77,8 +77,20 @@ public Member(
 
-    public void matchPassword(String password) {
-        this.password.validateMatches(password);
+    public void updateName(String name) {
+        this.name = name;
+    }
+
+    public void updateGender(String gender) {
+        this.gender = gender;
+    }
```
p3: 토끼 좋은데요~ 프론트엔드에서 검증하겠지만 백엔드에서 한 번 더 검증해 줘도 좋을 것 같습니다.

dye0p replied:
@coderabbitai
Patch 업데이트라 수정하지 않는 필드는 null로 들어와서 null 예외를 던지면 안될 것 같아. 그래도 null 검사를 해야할까?

dye0p replied:
@coderabbitai

RequestDTO에서 blank, 공백에 대한 검증을 하고 있는데, 서비스 레이어에서 또 해야 할까?

kimdevv reviewed (commented):
고생하셨습니다. 간단한 커멘트 남깁니다~

ppparkta reviewed (commented):
코드 잘 작성해 주셨네요. 토끼와 프리가 남겨준 코멘트 확인했는데 유효성 검사(특히 전화번호 중복검사)는 꼭 반영해야 할 것 같아요. 이 부분 반영 확인되면 바로 승인하겠습니다. 고생하셨습니다!

dye0p committed:
feat: 유효성 검증 예외 발생시 메시지 본문만 반환하도록 메세지 파싱 로직 추가

- MethodArgumentNotValidException 처리 시 첫 번째 필드 오류 메시지 반환
- 반환 메세지 형식은 "필드명 : 메세지" 형식임
- getRequestValidExceptionMessage 메서드 추가
- 기본 오류 메시지 '잘못된 요청입니다.' 처리 추가

dye0p committed:
feat: 성별 Enum 클래스 추가

dye0p committed:
refactor: 성별을 Enum 상수로 받도록 변경

dye0p committed:
feat: 회원정보 수정 전화번호 중복 검증 및 예외처리 추가

dye0p committed:
feat: 수정하려는 정보가 없는 경우 검증 및 예외처리 추가

dye0p committed:
test: 사용자 회원수정 E2E 테스트 추가

jye326 commented on backend/fittoring/src/main/java/fittoring/application/member/presentation/dto/request/MemberInfoUpdateRequest.java:
```diff
@@ -0,0 +1,20 @@
+
+public record MemberInfoUpdateRequest(
+        @Nullable
+        @Pattern(regexp = "^(?!\\s*$).+", message = "이름은 비어 있을 수 없습니다.")
```
p2: 값이 없거나 공백인 경우를 검사하는 것 같은데 복잡한 정규식 대신 `@NotBlank(message = "")`를 활용하면 좋을 것 같아요! `@NotBlank`는 null 검사도 포함하기 때문에 `MemberService.updateMemberInfo()`에서의 null 검사도 생략할 수 있겠네요!

dye0p replied:
이 부분은 정책상 수정하지 않는 정보는 null로 들어오는 것을 허용하기 위해서였습니다. `@NotBlank`로 null 검사를 해주면 서비스까지 요청이 가지 않기 때문에 사용할 수 없을 것 같습니다!

서비스에서 null 검사를 해주는 이유는 null로 들어온 필드에 대해서는 업데이트를 하지 않게하기 위해서였습니다!

jye326 replied:
그랬군요! 이해했습니다!

jye326 commented on backend/fittoring/src/main/java/fittoring/application/member/service/MemberService.java:
```diff
@@ -49,18 +52,56 @@ private Image findMentoringImage(Mentoring mentoring) {
+                .orElseThrow(() -> new MemberNotFoundException(BusinessErrorMessage.MEMBER_NOT_FOUND.getMessage()));
+    }
+
+    private void validateDuplicatePhoneNumber(MemberInfoUpdateRequest request) {
```
p2: `AuthService`에 이미 동일한 함수가 있으니 이를 `public`으로 활용하는 것은 어떨까요? 굳이 객체 생성 없이 사용할 수 있게 `public static`으로 사용하면 좋겠네요

dye0p replied:
static으로 선언하게되면 내부에 memberRepository로 static으로 선언되어야 하는데, static은 빈으로 등록할 수 없어서 static 선언은 힘들 것 같습니다..! 

중복이 심하다고 생각이 들진 않아서, 현재 방식을 유지하는 방법이 있고,
`PhoneValidator` 같은 컴포넌트 객체를 만들어 사용하는 방법이 있을 것 같아요. 

더 좋은 방법이 있다면 의견 남겨주세요!

jye326 replied:
Auth와 Member라는 도메인이 연관성이 커서 발생한 중복 같아요. 아직은 그대로 둬도 되겠지만 AuthService에서 MemberService의 기능을 사용하는 쪽으로 리팩터링 해도 될 것 같네요~

jye326 commented on backend/fittoring/src/test/java/fittoring/application/member/service/MemberServiceTest.java:
```diff
@@ -40,7 +49,7 @@ void successGetMyInfoForMentee() {
 
         // then
-        SoftAssertions.assertSoftly(softAssertions -> {
+        assertSoftly(softAssertions -> {
```
p3: 주체 파악을 위해 SoftAssertions를 static import 하지 않기로 정했던 것 같아요!

dye0p replied:
감사합니다 수정했습니다 👍

jye326 reviewed (commented):
수고하셨습니다!! 간단한 리뷰 남겼으니 확인 부탁드려요~!

dye0p committed:
test: 주체 파악을 위해 SoftAssertions의 static import 제거

jye326 reviewed (approved):
승인하겠습니다~
